### PR TITLE
Issue#9845.fix Binding tables using 'join .. on .. and column_x=?' or 'join .. on .. and cloumn_x in (?)' can not sharding correct.

### LINKED ISSUE #9845: Binding tables doesn't work when using 'join and in ()' condition
**Here is dependency**
<dependency>
	<groupId>org.apache.shardingsphere</groupId>
	<artifactId>sharding-jdbc-spring-boot-starter</artifactId>
	<version>4.1.0</version>
</dependency>

**Here is the output of my configuration**
bindingTables:
- t_order
- t_order_item
tables:
  t_order:
    actualDataNodes: ds0.t_order_${0..1}
    keyGenerator:
      column: order_id
      type: SNOWFLAKE
    logicTable: t_order
    tableStrategy:
      inline:
        algorithmExpression: t_order_${order_id % 2}
        shardingColumn: order_id
  t_order_item:
    actualDataNodes: ds0.t_order_item_${0..1}
    logicTable: t_order_item
    tableStrategy:
      inline:
        algorithmExpression: t_order_item_${order_id % 2}
        shardingColumn: order_id

**Here is part of log:**
result:
       Logic SQL: SELECT
        o.order_id, o.user_id, o.order_name, i.item_name, i.price, i.quantity
        FROM t_order o
        JOIN t_order_item i
        on o.order_id = i.order_id
        where o.order_id in (?)
      Actual SQL: ds0 ::: SELECT
        o.order_id, o.user_id, o.order_name, i.item_name, i.price, i.quantity
        FROM t_order_1 o
        JOIN t_order_item_1 i
        on o.order_id = i.order_id
        where o.order_id in (?) ::: [1]
**another result when I use 'and' instead of 'where'：**
     Logic SQL: SELECT
            o.order_id, o.user_id, o.order_name, i.item_name, i.price, i.quantity
            FROM t_order o
            JOIN t_order_item i
            on o.order_id = i.order_id
            and o.order_id in (?)
    Actual SQL: ds0 ::: SELECT
            o.order_id, o.user_id, o.order_name, i.item_name, i.price, i.quantity
            FROM t_order_0 o
            JOIN t_order_item_1 i
            on o.order_id = i.order_id
            and o.order_id in (?) ::: [1]
    Actual SQL: ds0 ::: SELECT
            o.order_id, o.user_id, o.order_name, i.item_name, i.price, i.quantity
            FROM t_order_0 o
            JOIN t_order_item_0 i
            on o.order_id = i.order_id
            and o.order_id in (?) ::: [1]
    Actual SQL: ds0 ::: SELECT
            o.order_id, o.user_id, o.order_name, i.item_name, i.price, i.quantity
            FROM t_order_1 o
            JOIN t_order_item_1 i
            on o.order_id = i.order_id
            and o.order_id in (?) ::: [1]
    Actual SQL: ds0 ::: SELECT
            o.order_id, o.user_id, o.order_name, i.item_name, i.price, i.quantity
            FROM t_order_1 o
            JOIN t_order_item_0 i
            on o.order_id = i.order_id
            and o.order_id in (?) ::: [1]
       
As you can see, different results happened

I tried to debug, I found the name of  'org.apache.shardingsphere.sharding.route.engine.ShardingRouteDecorator#getShardingConditions' . line 82.
Maybe that's the problem when I using 'and'.  WhereClauseShardingConditionEngine is the only implementation for Sharding-JDBC. 
Is there any necessary to add a class to support 'and in' condition?

### PR DESCRIPTION
**Pull request by huanghao495430759**
Fixes #9845 .

Changes proposed in this pull request:
- Binding tables using 'join .. on .. and column_x=?' or 'join .. on .. and cloumn_x in (?)'  sharding correct.

### CODE REVIEW

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/util/WhereSegmentExtractUtils.java [hunk 1]
```diff
@@ -41,7 +41,28 @@
...
+            return Collections.emptyList();
+        }
+        TableSegment tableSegment = selectStatement.getFrom();
+        Collection<WhereSegment> result = new LinkedList<>();
+        if (tableSegment instanceof JoinTableSegment && null != ((JoinTableSegment) tableSegment).getCondition()) {
+            ExpressionSegment expressionSegment = ((JoinTableSegment) tableSegment).getCondition();
+            WhereSegment whereSegment = new WhereSegment(expressionSegment.getStartIndex(), expressionSegment.getStopIndex(), expressionSegment);
+            result.add(whereSegment);
+        }
+
```
Comment (tristaZero) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/util/WhereSegmentExtractUtils.java:62]: Please remove this blank line.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-rewrite/src/test/resources/sharding/select.xml [hunk 2]
```diff
@@ -453,4 +453,19 @@
         <input sql="SELECT * FROM t_account, t_account_detail WHERE t_account.account_id = 1 AND t_account.account_id = t_account_detail.account_id" />
         <output sql="SELECT * FROM t_account_1, t_account_detail_1 WHERE t_account_1.account_id = 1 AND t_account_1.account_id = t_account_detail_1.account_id" />
     </rewrite-assertion>
+
+    <rewrite-assertion id="select_with_join_table_on_condition_with_qualified_columns">
```
Comment (tristaZero) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-rewrite/src/test/resources/sharding/select.xml:457]: Hi, this PR looks generally correct,  thanks for your work.
Also, the focus of the relevant issue is about `whereSegment` and SQL route, right? So could you add an integrated test to make sure this `join and in` SQL has a correct execution result completely?

### How?
Please refer to the case in line 102 of `shardingsphere-test/shardingsphere-integration-test/shardingsphere-integration-test-suite/src/test/resources/cases/dql/dql-integration-test-cases.xml`. This case is very similar to your PR.

And last but not least, have a great holiday, :)

**Review by tristaZero — CHANGES_REQUESTED** [review]

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-rewrite/src/test/resources/sharding/select.xml (hunk 2 above)
Comment (huanghao495430759) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-rewrite/src/test/resources/sharding/select.xml:457]: OK. I will add any case for this PR.

**Review by tristaZero — CHANGES_REQUESTED** [review]
Hi is there any progress?

File: shardingsphere-test/shardingsphere-integration-test/shardingsphere-integration-test-suite/src/test/resources/cases/dql/dql-integration-test-cases.xml [hunk 3]
```diff
@@ -19,378 +19,387 @@
     <test-case sql="SELECT 1 as a">
         <assertion expected-data-file="select_constant_without_table.xml" />
     </test-case>
-
```
Comment (tristaZero) [inline shardingsphere-test/shardingsphere-integration-test/shardingsphere-integration-test-suite/src/test/resources/cases/dql/dql-integration-test-cases.xml:22]: Unit tests look good to me, but why are many lines modified?

**Review by tristaZero — CHANGES_REQUESTED** [review]

**Review by tristaZero — APPROVED** [review]
Oh, these changes look great and concise.

**Pull request merged**